Reduce unsafeness in StreamClientConnection and WebViewImpl

annevk opened the pull request:
#### 233a9daa785641cbe7ecf39e768f8318b31c7335
<pre>
Reduce unsafeness in StreamClientConnection and WebViewImpl
<a href="https://bugs.webkit.org/show_bug.cgi?id=295486">https://bugs.webkit.org/show_bug.cgi?id=295486</a>

Reviewed by Chris Dumez and Rupin Mittal.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):

m_connection is a const Ref and therefore does not need stack protection.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:

All equivalent interfaces with member pointing to a WebKit::WebViewImpl
already have it as WeakPtr so it seems these were an oversight.

Canonical link: <a href="https://commits.webkit.org/297113@main">https://commits.webkit.org/297113@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d0603e37de18fffbd66e59fee851afcc9964aca

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110563 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30222 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20655 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116589 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60830 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38810 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84079 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113511 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99557 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64520 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24037 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60384 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94059 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17756 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119379 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37603 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27921 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93043 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37977 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95828 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92866 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37877 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15617 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33576 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17840 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37498 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42969 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37161 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40500 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38869 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -139,8 +139,10 @@ class StreamClientConnection final : public ThreadSafeRefCounted<StreamClientCon
+
         CheckedRef<StreamClientConnection> m_owner;
-        Connection::Client& m_receiver;
+        WeakPtr<Connection::Client> m_receiver;
```
`WeakRef` since `m_receiver` cannot be null.

RupinMittal commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -139,8 +139,10 @@ class StreamClientConnection final : public ThreadSafeRefCounted<StreamClientCon
         void didClose(Connection&) final;
         void didReceiveInvalidMessage(Connection&, MessageName, int32_t indexOfObjectFailingDecoding) final;
     private:
+        RefPtr<Connection::Client> protectedReceiver() { return m_receiver.get(); }
```
Once `m_receiver` is changed to be a `WeakRef`, this function should return a `Ref`.

RupinMittal commented on Source/WebKit/Platform/IPC/StreamClientConnection.cpp:
```diff
@@ -114,7 +114,7 @@ void StreamClientConnection::setMaxBatchSize(unsigned size)
 {
     m_dedicatedConnectionClient.emplace(*this, receiver);
-    m_connection->open(*m_dedicatedConnectionClient, dispatcher);
+    m_connection->open(CheckedRef { *m_dedicatedConnectionClient }.get(), dispatcher);
```
We can use `Ref` here instead of `CheckedRef`.

RupinMittal reviewed (commented):
A few comments. Also, since you've fixed all the failures in these files, don't forget to enjoy the fruits of your labor and update the expectations :)

RupinMittal reviewed (approved):
LGTM

cdumez commented on Source/WebKit/Platform/IPC/StreamClientConnection.h:
```diff
@@ -139,8 +139,10 @@ class StreamClientConnection final : public ThreadSafeRefCounted<StreamClientCon
+
         CheckedRef<StreamClientConnection> m_owner;
-        Connection::Client& m_receiver;
+        WeakRef<Connection::Client> m_receiver;
```
We could use `const CheckedRef<>` here I believe and then we wouldn't need `protectedReceiver()` no?

RupinMittal replied:
Good point! Initially, this patch was changing the raw reference to `WeakPtr`, so I recommended `WeakRef`. I didn't suggest making it `const` since it's `Weak`, so it wouldn't allow us to eliminate `protectedReceiver()`. But `const CheckedRef` should.

@annevk This is the right change to make here.

annevk committed:
Reduce unsafeness in StreamClientConnection and WebViewImpl
https://bugs.webkit.org/show_bug.cgi?id=295486

Reviewed by Chris Dumez and Rupin Mittal.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):

m_connection is a const Ref and therefore does not need stack protection.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:

All equivalent interfaces with member pointing to a WebKit::WebViewImpl
already have it as WeakPtr so it seems these were an oversight.

Canonical link: https://commits.webkit.org/297113@main
